crates/apollo-compiler/src/database/ast.rs

```diff
@@ -1,20 +1,26 @@
 use apollo_parser::{Parser as ApolloParser, SyntaxTree};
 use rowan::GreenNode;
 
-use crate::{database::inputs::InputDatabase, diagnostics::SyntaxError, ApolloDiagnostic};
+use crate::{database::inputs::InputDatabase, diagnostics::SyntaxError, ApolloDiagnostic, FileId};
 
 #[salsa::query_group(AstStorage)]
 pub trait AstDatabase: InputDatabase {
-    fn ast(&self) -> SyntaxTree;
+    /// Get an AST for a particular file. Returns a `rowan` SyntaxTree.  The
+    /// SyntaxTree can be safely shared between threads as it's `Send` and
+    /// `Sync`.
+    fn ast(&self, file_id: FileId) -> SyntaxTree;
 
-    // root node
-    fn document(&self) -> GreenNode;
+    /// Get a file's GraphQL Document. Returns a `rowan` Green Node. This is the
+    /// top level document node that can be used when going between an
+    /// SyntaxNodePtr to an actual SyntaxNode.
+    fn document(&self, file_id: FileId) -> GreenNode;
 
+    /// Get syntax errors found in the compiler's manifest.
     fn syntax_errors(&self) -> Vec<ApolloDiagnostic>;
 }
 
-fn ast(db: &dyn AstDatabase) -> SyntaxTree {
-    let input = db.input();
+fn ast(db: &dyn AstDatabase, file_id: FileId) -> SyntaxTree {
```

Comment:
The AST will always be computed in the same way for schemas or queries, since they are part of the same GraphQL Document grammar. But to ensure that the incoming queries (especially for the router) do not rewrite the schema context, we need to think of a way to separate _what_ can be added as HIR nodes from a query (only Operation and Fragment definitions). We should likely have a change to the HIR structure where `Definition` is indeed an enum of two types `ExecutableDefinition` and `TypeSystemDefinitionOrExtension`. Maybe then this function can distinguish between the two?